src/librustdoc/clean/mod.rs

```diff
@@ -1212,14 +1213,19 @@ fn resolve(cx: &DocContext, path_str: &str, is_val: bool) -> Result<(Def, Option
                         Def::Enum(_) => true,
                         _ => false,
                     };
-                    let elem = if is_enum {
-                        cx.tcx.adt_def(did).all_fields().find(|item| item.ident.name == item_name)
-                    } else {
-                        cx.tcx.adt_def(did)
-                              .non_enum_variant()
-                              .fields
-                              .iter()
-                              .find(|item| item.ident.name == item_name)
+                    let elem = match cx.tcx.type_of(did).sty {
+                        ty::TyAdt(def, _) => {
+                            if is_enum {
```

Comment:
You don't even need to switch on `is_enum` here: `def.all_fields()` should work in general.